src/main/java/com/powsybl/balances_adjustment/util/ControlArea.java

```diff
@@ -0,0 +1,132 @@
+        Map<VoltageLevel, Integer> vlNums = new HashMap<>();
+        UndirectedGraph<Object, Object> voltageLevelGraph = createVoltageLevelGraph(network, vlNums);
+        for (Object o : terminalsAndBoundaries) {
+            voltageLevelGraph.traverse(vlNums.get(getVoltageLevel(o)), (v1, e, v2) -> findContainedVoltageLevels(voltageLevelGraph.getEdgeObject(e),
+                    voltageLevelIds, terminalsAndBoundaries));
+        }
+        return voltageLevelIds;
+    }
+
+    private static UndirectedGraph<Object, Object> createVoltageLevelGraph(Network network, Map<VoltageLevel, Integer> vlNums) {
+        UndirectedGraph<Object, Object> voltageLevelGraph = new UndirectedGraphImpl<>();
+        network.getVoltageLevelStream().forEach(vl -> {
+            int v = voltageLevelGraph.addVertex();
+            vlNums.put(vl, v);
+        });
+        network.getBranchStream().forEach(b -> voltageLevelGraph.addEdge(vlNums.get(b.getTerminal1().getVoltageLevel()), vlNums.get(b.getTerminal2().getVoltageLevel()), b));
+        network.getThreeWindingsTransformerStream().forEach(twt -> {
+            int v = voltageLevelGraph.addVertex();
+            voltageLevelGraph.addEdge(vlNums.get(twt.getLeg1().getTerminal().getVoltageLevel()), v, twt.getLeg1());
+            voltageLevelGraph.addEdge(v, vlNums.get(twt.getLeg2().getTerminal().getVoltageLevel()), twt.getLeg2());
+            voltageLevelGraph.addEdge(v, vlNums.get(twt.getLeg3().getTerminal().getVoltageLevel()), twt.getLeg3());
+        });
+        network.getDanglingLineStream().forEach(dl -> {
+            int v = voltageLevelGraph.addVertex();
+            voltageLevelGraph.addEdge(vlNums.get(dl.getTerminal().getVoltageLevel()), v, dl);
+        });
+        return voltageLevelGraph;
+    }
+
+    private static VoltageLevel getVoltageLevel(Object o) {
+        if (o instanceof Terminal) {
+            return ((Terminal) o).getVoltageLevel();
+        } else if (o instanceof Boundary) {
+            return ((Boundary) o).getVoltageLevel();
+        } else {
+            throw new AssertionError();
+        }
+    }
+
+    private static TraverseResult findContainedVoltageLevels(Object edge, Set<String> voltageLevelIds, Set<Object> terminalsAndBoundaries) {
+        if (edge instanceof Branch) {
+            Branch branch = (Branch) edge;
+            voltageLevelIds.add(branch.getTerminal1().getVoltageLevel().getId());
+            if (containsTieFlows(branch, Branch.Side.ONE, terminalsAndBoundaries)) {
+                return TraverseResult.TERMINATE;
+            }
+            voltageLevelIds.add(branch.getTerminal2().getVoltageLevel().getId());
+            if (containsTieFlows(branch, Branch.Side.TWO, terminalsAndBoundaries)) {
+                return TraverseResult.TERMINATE;
+            }
+        } else if (edge instanceof ThreeWindingsTransformer.Leg) {
+            ThreeWindingsTransformer.Leg leg = (ThreeWindingsTransformer.Leg) edge;
```

Comment:
Not sure because:
- in 3WT, the star bus (the fictitious middle bus) is never in `terminalAndBoundaries` so if we are from the star bus, then the leg's voltage level is necessary contained. If we are from the leg's voltage level (to the star bus), the leg's voltage level is contained too so the behavior is the same, isn't it?
- in DL, if we come from the fictitious boundary's voltage level, the dangling line's voltage level is necessary contained. If we are from the dangling line's voltage level, it is contained too.